Update documentation for OpenAPI in minimal APIs (#27181)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/minimal-apis/openapi.md

* `UseSwagger`adds the [Swagger middleware](xref:tutorials/get-started-with-swashbuckle#add-and-configure-swagger-middleware).
* `UseSwaggerUI` enables the [Static File Middleware](xref:fundamentals/static-files).
* <xref:Microsoft.AspNetCore.Builder.RoutingEndpointConventionBuilderExtensions.WithName%2A>: The <xref:Microsoft.AspNetCore.Routing.IEndpointNameMetadata> on the endpoint is used for link generation and is treated as the operation ID in the given endpoint's OpenAPI specification.

> * [WithOpenApi](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) is explained later in this article.

Reviewer: Is this missing the code quotes to match the bullets above?

```suggestion
* [`WithOpenApi`](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) is explained later in this article.
```

When using [`Swashbuckle.AspNetCore`](https://www.nuget.org/packages/Swashbuckle.AspNetCore/) with `Microsoft.AspNetCore.OpenApi`, `Swashbuckle.AspNetCore` 6.3.1 and later must be used. Version 6.4.0 or later is required if you want to [provide endpoint summary or description text](#add-endpoint-summary-or-description).

### Add OpenAPI annotations to endpoints via `WithOpenApi`

> Calling [WithOpenApi](xref:Microsoft.Aspnetcore.Builder.OpenApiEndpointConventionBuilderExtensions.WithOpenApi) to the endpoints metadata. This metadata can be:

Reviewer 2: ```suggestion
Calling [WithOpenApi](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) to the endpoints metadata. This metadata can be:
```

* Consumed in third-party packages like [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/).
* Displayed in the Swagger user interface or in YAML or JSON generated to define the file.

[!code-csharp[](7.0-samples/todo/Program.cs?name=snippet_withopenapi&highlight=9)]

#### Modify the OpenAPI annotation in `WithOpenApi`

> The [WithOpenApi](xref:Microsoft.Aspnetcore.Builder.OpenApiEndpointConventionBuilderExtensions.WithOpenApi) method accepts a function that can be used to modify the OpenAPI annotation. For example, in the following code, a description is added to the first parameter of the endpoint:

Reviewer 2: ```suggestion
The [WithOpenApi](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) method accepts a function that can be used to modify the OpenAPI annotation. For example, in the following code, a description is added to the first parameter of the endpoint:
```

When using [`Swashbuckle.AspNetCore`](https://www.nuget.org/packages/Swashbuckle.AspNetCore/) with `Microsoft.AspNetCore.OpenApi`, `Swashbuckle.AspNetCore` 6.3.1 and later must be used. Version 6.4.0 or later is required if you want to [provide endpoint summary or description text](#add-endpoint-summary-or-description).

### Add OpenAPI annotations to endpoints via `WithOpenApi`

> Calling [WithOpenApi](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi)to the endpoints metadata. This metadata can be:

Reviewer 3: Missing something, "Calling WithOpenApi to the endpoints metadata"?

> Calling [WithOpenApi](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) on the endpoint to the endpoints metadata. This metadata can be:

Reviewer: ```suggestion
Calling [`WithOpenApi`](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) on the endpoint to the endpoints metadata. This metadata can be:
```

* Consumed in third-party packages like [Swashbuckle.AspNetCore](https://www.nuget.org/packages/Swashbuckle.AspNetCore/).

> * Displayed in the Swagger user interface or in YAML or JSON generated to define the file.

Reviewer: ```suggestion
* Displayed in the Swagger user interface or in YAML or JSON generated to define the API.
```

[!code-csharp[](7.0-samples/todo/Program.cs?name=snippet_withopenapi&highlight=9)]

#### Modify the OpenAPI annotation in `WithOpenApi`

> The [WithOpenApi](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) method accepts a function that can be used to directly modify the OpenAPI annotation. For example, in the following code, a description is added to the first parameter of the endpoint:

Reviewer: Given the method actually needs to return an `OpenApiOperation` instance, I'm not sure about the "modify" in this statement, but might be simpler to just leave it this way for now.
```suggestion
The [`WithOpenApi`](/dotnet/api/microsoft.aspnetcore.builder.openapiendpointconventionbuilderextensions.withopenapi) method accepts a function that can be used to modify the OpenAPI annotation. For example, in the following code, a description is added to the first parameter of the endpoint:
```

[!code-csharp[](7.0-samples/todo/Program.cs?name=snippet_withopenapi2&highlight=9-99)]

> ### Add operation IDs to Open API

Reviewer: ```suggestion
### Add operation IDs to OpenAPI
```

> Operation IDs are used to uniquely identify a given endpoints in OpenAPI. The `WithName` extension method can be used to set the operation ID used for a method.

Reviewer 3: ```suggestion
Operation IDs are used to uniquely identify a given endpoint in OpenAPI. The `WithName` extension method can be used to set the operation ID used for a method.
```

Reviewer 3: Link to https://learn.microsoft.com/dotnet/api/microsoft.aspnetcore.builder.routingendpointconventionbuilderextensions.withname?

> Operation IDs are used to uniquely identify a given endpoint in OpenAPI. The [`WithName`](/dotnet/api/microsoft.aspnetcore.builder.routingendpointconventionbuilderextensions.withname) extension method can be used to set the operation ID used for a method.

Reviewer: Do you think there's value in linking to the spec docs? e.g.

```suggestion
Operation IDs are used to uniquely identify a given endpoint in OpenAPI. The [`WithName`](/dotnet/api/microsoft.aspnetcore.builder.routingendpointconventionbuilderextensions.withname) extension method can be used to set the [operationId](https://spec.openapis.org/oas/latest.html#fixed-fields-7) used for a method.
```

Author: Linking to the spec is really hard. Ideally, I think linking to the `Microsoft.OpenApi` docs would provide the most tangible value for users ("oh this is the type I need to pass in"). I filed an issue to that effect over at https://github.com/microsoft/OpenAPI.NET/issues/1026.

Reviewer: Fair enough. We should make a call on how to refer to the fields then, i.e. verbatim like "operationId", or in descriptive fashion like "operation ID". Personally I prefer the former as it matches what is being set and how it appears in the produced document, appears in spec docs, etc.

Author: TBH, I prefer the later because there is a disparity between how the operation ID is set in the OpenAPI.NET models (**O**perationId) versus how it appears in the serialized document and the UI (**o**operationId) so I think the descriptive reference is the most implementation neutral. Also the code sample already shows how to set it.

[!code-csharp[](samples/todo/Program.cs?name=snippet_name)]

Alternatively, the `OperationId` property can be set directly on the OpenApi annotation.

```csharp
app.MapGet("/todos", async (TodoDb db) => await db.Todos.ToListAsync())
    .WithOpenApi(operation => new(operation)
    {
        OperationId = "GetTodos"
    });
```

> ### Add tags to the Open API description

Reviewer: ```suggestion
### Add tags to the OpenAPI description
```

> OpenAPI supports using [tag objects](https://swagger.io/docs/specification/grouping-operations-with-tags/) to categorize operations. These tags are typically used to group operations in the Swagger UI. These tags can be added to an operation by invoking the [WithTags](/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.withtags) extension method on the endpoint with the desired tags.

Reviewer: Should we link to the OpenAPI spec instead of swagger.io perhaps?

Author: The Swagger docs are a little bit more helpful at informing the user _how_ tags should be used instead of just telling users what they are so they feel more helpful to link out to.

Reviewer: OK. I admit something about linking to a commercial site (that isn't ms.com) turns me off.

[!code-csharp[](samples/todo/Program.cs?name=snippet_grp)]

> Alternatively, the list of `OpenApiTags` can be set on the OpenAPI annotation via the `WithOpenApi` extension method. d

Reviewer 3: ```suggestion
Alternatively, the list of `OpenApiTags` can be set on the OpenAPI annotation via the `WithOpenApi` extension method.
```

```csharp
app.MapGet("/todos", async (TodoDb db) => await db.Todos.ToListAsync())
    .WithOpenApi(operation => new(operation)
    {

>         Tags = new List<OpenApiTag>() { new() { Name = "Todos" } }

Reviewer: ```suggestion
        Tags = new List<OpenApiTag> { new() { Name = "Todos" } }
```

    });
```

### Add endpoint summary or description

The following highlighted code provides summary and description text for OpenAPI spec generation. You can call extension methods or use attributes:

:::code language="csharp" source="endpoint-description-summary/7.0-samples/OpenAPISummary/Program.cs" id="summary_and_description" highlight="13-14":::

:::code language="csharp" source="endpoint-description-summary/7.0-samples/OpenAPISummary/Program.cs" id="summary_and_description2" highlight="2-3":::

Alternatively, the summaries can be set directly on the OpenAPI annotation.

```csharp
app.MapGet("/todoitems2", async (TodoDb db) => await db.Todos.ToListAsync())
    .WithOpenApi(operation => new(operation)
    {
        Summary = "This is a summary",
        Description = "This is a description"
    });
```

### Exclude Open API description

In the following sample, the `/skipme` endpoint is excluded from generating an OpenAPI description:

[!code-csharp[](7.0-samples/WebMinAPIs/Program.cs?name=snippet_swag2&highlight=20-21)]

### Mark an API as obsolete

> To mark an endpoint as obsolete, set the `Depercated` property on the OpenAPI annotation.

Reviewer 3: ```suggestion
To mark an endpoint as obsolete, set the `Deprecated` property on the OpenAPI annotation.
```

```csharp
app.MapGet("/todos", async (TodoDb db) => await db.Todos.ToListAsync())
    .WithOpenApi(operation => new(operation)
    {
        Deprecated = true
    });
```

### Describe response types

> OpenAPI supports providing an description of the responses returned from an API. Minimal APIs supports three strategies for setting the response type of an endpoint:

Reviewer 3: ```suggestion
OpenAPI supports providing a description of the responses returned from an API. Minimal APIs supports three strategies for setting the response type of an endpoint:
```

* Via the `Produces` extension method
* Via the `ProducesResponseType` attribute

> * By returning `TypedResults` from your application

Reviewer: ```suggestion
* By returning `TypedResults` from your route handlers
* By returning `IResult` types that implement `IEndpointMetadataProvider` from your route handlers
```

The `Produces` extension method can be used to add `Produces` metadata to an endpoint. When no parameters are provided, the extension method will populate metadata for the targeted type under a `200` status code and an `application/json` content type.

```csharp
app
    .MapGet("/todos", async (TodoDb db) => await db.Todos.ToListAsync())
    .Produces<IList<Todo>>();
```

Leveraging `TypedResults` in the implementation of an endpoint's route handler will automatically include the response type metadata for the endpoint. For example, the code below will automatically annotate the endpoint with a response under the `200` status code with an `application/json` content type.

```csharp
app.MapGet("/todos", async (TodoDb db) =>
{
    var todos = await db.Todos.ToListAsync());
    return TypedResults.Ok(todos);
});
```

> #### Set responses for ProblemDetails

Reviewer: ```suggestion
#### Set responses for `ProblemDetails`
```

> When setting the response type for endpoints that may return a ProblemDetails response, the `ProducesProblem` extension method can be used to provide the annotation or the `TypedResults.Problem` return type.

Reviewer 3: suggestion:
"the `ProducesProblem` extension method or the `TypedResults.Problem` return type can be used to provide the annotation."

> When setting the response type for endpoints that may return a ProblemDetails response, the [`ProducesProblem`](/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.producesproblem) extension method or [`TypeResults.Problem`](/dotnet/api/microsoft.aspnetcore.http.typedresults.problem?) can be used to add the appropriate annotation to the endpoint's metadata.

Reviewer: ```suggestion
When setting the response type for endpoints that may return a `ProblemDetails` response, the [`ProducesProblem`](/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.producesproblem) extension method or [`TypedResults.Problem`](/dotnet/api/microsoft.aspnetcore.http.typedresults.problem?) can be used to add the appropriate annotation to the endpoint's metadata.
```

When there are no explicit annotations provided by one of the strategies above, the framework will attempt to determine a default response type by examining the signature of the response. This default response will be populated under the `200` status code in the OpenAPI definition.

### Describe request body and parameters

In addition to describing the types that are returned by an endpoint, OpenAPI also supports annotating the inputs that are consumed by an API. These inputs fall into two categories:

* Parameters that appear in the path, query string, headers, or cookies
* Data transmitted as part of the request body

The framework will infer the types for request parameters in the path, query, and header string automatically based on the signature of the route handler.

To define the type of inputs transmitted as the request body, you can configure the properties by using the `Accepts` extension method to define the object type and content type that are expected by the request handler.

```csharp
app.MapPost("/todos/{id}", (int id, Todo todo) => ...)
  .Accepts<Todo>();
```

> When there is no explicit annotation provide, the framework will attempt to determine the default request type if there exists a request body parameter in the endpoint handler. 

Reviewer: This might be clearer with an example, e.g. the previously shown example for `Accepts` isn't actually required given that accepting the `Todo` parameter (assuming it doesn't implement `BindAsync`) will default to inferring accepts metadata of that shape and application/json. Also consider mentioning `IEndpointParameterMetadata` as an additional strategy that's used to acquire accepts metadata.

### Describe request body and parameters

In addition to describing the types that are returned by an endpoint, OpenAPI also supports annotating the inputs that are consumed by an API. These inputs fall into two categories:

* Parameters that appear in the path, query string, headers, or cookies
* Data transmitted as part of the request body

The framework infers the types for request parameters in the path, query, and header string automatically based on the signature of the route handler.

To define the type of inputs transmitted as the request body, configure the properties by using the [`Accepts`](/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.accepts) extension method to define the object type and content type that are expected by the request handler. In the example below, the endpoint accepts a `Todo` object in the request body with an expected content-type of `application/xml`.

```csharp
app.MapPost("/todos/{id}", (int id, Todo todo) => ...)
  .Accepts<Todo>("application/xml");
```

> In addition the [`Accepts`](/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.accepts) extension method, it is possible for a parameter type to describe its own annotation by implementing the [`IEndpointParameterMetadataProvider`](/dotnet/api/microsoft.aspnetcore.http.metadata.iendpointparametermetadataprovider) interface. For example, the following `Todo` type adds an annotation that requires a request body with an `application/xml` content-type. 

Reviewer 4: ```suggestion
In addition to the [`Accepts`](/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.accepts) extension method, it is possible for a parameter type to describe its own annotation by implementing the [`IEndpointParameterMetadataProvider`](/dotnet/api/microsoft.aspnetcore.http.metadata.iendpointparametermetadataprovider) interface. For example, the following `Todo` type adds an annotation that requires a request body with an `application/xml` content-type. 
```
